### PR TITLE
feat: Add user_dn_hash field to improve User::getFromDBbyDn performance

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !30554

**Description:**

This PR introduces a new field `user_dn_hash` in the `glpi_users` table. This field is a virtual field that is generated by hashing the `user_dn` field. This change is aimed at improving the performance of queries that involve the `user_dn` field by using the hashed value instead.

**Changes:**

1. Added `ext-hash` to the required PHP extensions in `composer.json` and `composer.lock` to avoid soft dependency issues.
2. Created a new migration file `user.php` in `install/migrations/update_10.0.10_to_10.0.11/` to add the `user_dn_hash` field and its index to the `glpi_users` table.
3. Updated `glpi-empty.sql` to include the `user_dn_hash` field and its index in the `glpi_users` table creation statement.
4. Updated the `getFromDBbyDn` method in `User.php` to use the `user_dn_hash` field for the database query.
5. Updated the test file `AuthLdap.php` to unset the `user_dn_hash` field in the `dup` array.

### CODE REVIEW

**Comment by trasher** [comment]
Are we sure a new PHP extension is really to be required? Anyway, ot should not be in bugufxes; and not only in composer.json

**Comment by AdrienClairembault** [comment]
I didn't realize `sha256` would require a new PHP extension.

We could use `md5` instead to avoid this but could it be exploited by malicious users ?

**Comment by trasher** [comment]
> We could use `md5` instead to avoid this but could it be exploited by malicious users ?

I'm not sure is can really be used for an  exploit; we can also use `sha1` which is a bit better maybe.

We rely on some `hash` features, but in main only (`hash_hmac`)

**Comment by AdrienClairembault** [comment]
> Anyway, it should not be in bugfixes;

Regarding the bugfixes/main debate, this change is introduced because the current `user_dn` column can't be indexed and it block the ldap sync for some users with big databases (as it take too long to complete).

So IMO this is kinda a bugfix and it would be best to stay on the 10.0.11 target (as long as we can drop the new ext requirement).

**Comment by AdrienClairembault** [comment]
Regarding the issues: 

1\) Composer hash

It seems @cedric-anne already encountered the issue and fixed in directly in composer require checker : https://github.com/maglnet/ComposerRequireChecker/pull/399

Maybe it was not fixed completely and we should open an issue on their side ?

2\) Unsupported index on virtual fields for Mysql + Myisam 

Should we fallback to a "real" field or can we make an exception for this ?
It would suck not being able to use a virtual fields because an obsolete engine do not support indexes on them.

**Comment by trasher** [comment]
> It would suck not being able to use a virtual fields because an obsolete engine do not support indexes on them.

Switching to InnoDB is not yet mandatory, we cannot change this is a bugfixes version.

**Comment by AdrienClairembault** [comment]
> > It would suck not being able to use a virtual fields because an obsolete engine do not support indexes on them.
> 
> Switching to InnoDB is not yet mandatory, we cannot change this is a bugfixes version.

Yeah I expect that of course, but maybe the error should not be flagged as an issue ? It wont make anything worse for users on myisam while it will improve speed for anyone else.

(Can we still drop myisam for GLPI 10.1 tho ?)

File: src/User.php [hunk 1]
```diff
@@ -558,7 +558,9 @@ public function getFromDBbySyncField($value)
      */
     public function getFromDBbyDn($user_dn)
     {
-        return $this->getFromDBByCrit(['user_dn' => $user_dn]);
+        return $this->getFromDBByCrit([
+            'user_dn_hash' => hash('sha256', $user_dn)
```
Comment (cedric-anne) [inline src/User.php:562]: Why not using a simple `sha1($user_dn)` ? We just want to have a smaller indexable value here, and we store the clear value in the `user_dn` field, so there is no need to use a secured hashing algo here. If you use `sha1()` or `md5()`, you will be able to remove the requirement of the `hash` extension that make `composer-require-checker` unhappy.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
I guess we'll have to stick to a normal field then (sadly).

@ccailly can you rework the field into a concrete field ?

**Comment by trasher** [comment]
Please target main branch, non new feature should reach 10.0/bf one.

**Comment by ccailly** [comment]
@trasher I rebased on main as you asked yesterday but I don't see your issue in the PR thread anymore. Should I keep main on target or switch back to bugfixes?

**Comment by trasher** [comment]
keep in main for now, once approved we'll see where it will go with @cedric-anne and @orthagh

File: src/User.php [hunk 2]
```diff
@@ -557,15 +557,15 @@ public function getFromDBbySyncField($value)
...
-     * @since 0.84
-     *
      * @param string $user_dn dn of the user
      *
      * @return boolean
      */
     public function getFromDBbyDn($user_dn)
     {
-        return $this->getFromDBByCrit(['user_dn' => $user_dn]);
+        return $this->getFromDBByCrit([
```
Comment (orthagh) [inline src/User.php:566]: maybe add a comment why getting the user by it's hash here.
The future maintener will thnak you

**Comment by trasher** [comment]
Please also add unit tests; this should not be too hard.

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: src/User.php [hunk 3]
```diff
@@ -890,6 +895,14 @@ public function post_addItem()
                 $right->add($affectation);
             }
         }
+
+        // Hash user_dn if set
+        if (isset($this->input['user_dn'])) {
```
Comment (cedric-anne) [inline src/User.php:900]: It could be preferable to no hash empty values.
```suggestion
        if (isset($this->input['user_dn']) && is_string($this->input['user_dn']) && strlen($this->input['user_dn']) > 0) {
```

File: install/migrations/update_10.0.11_to_10.0.12/user.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * ---------------------------------------------------------------------
+ */
+
+/**
+ * @var \DBmysql $DB
+ * @var \Migration $migration
+ */
+
+ // Add user_dn_hash field
+$migration->addField('glpi_users', 'user_dn_hash', 'varchar(32)', [
```
Comment (cedric-anne) [inline install/migrations/update_10.0.11_to_10.0.12/user.php:42]: I think a default empty value is safer than no default value. Indeed, creating a user without setting this field should not be a problem (when user is not connected to a LDAP).

```suggestion
$migration->addField('glpi_users', 'user_dn_hash', "varchar(32) NOT NULL DEFAULT ''", [
```

File: install/mysql/glpi-empty.sql [hunk 5]
```diff
@@ -7610,6 +7610,7 @@ CREATE TABLE `glpi_users` (
   `password_forget_token` char(40) DEFAULT NULL,
   `password_forget_token_date` timestamp NULL DEFAULT NULL,
   `user_dn` text,
+  `user_dn_hash` varchar(32),
```
Comment (cedric-anne) [inline install/mysql/glpi-empty.sql:7613]:
```suggestion
  `user_dn_hash` varchar(32) NOT NULL DEFAULT '',
```

File: src/User.php [hunk 6]
```diff
@@ -1185,6 +1198,14 @@ public function post_updateItem($history = true)
                 true
             );
         }
+
+        // Hash user_dn if is updated
+        if (in_array('user_dn', $this->updates)) {
+            $this->update([
+                'id' => $this->fields['id'],
+                'user_dn_hash' => md5($this->fields['user_dn'])
```
Comment (cedric-anne) [inline src/User.php:1206]: Can be updated to null, I guess.
```suggestion
                'user_dn_hash' => is_string($this->fields['user_dn']) && strlen($this->fields['user_dn']) > 0 ? md5($this->fields['user_dn']) : '',
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: install/migrations/update_10.0.11_to_10.0.12/user.php (hunk 4 above)
Comment (cedric-anne) [reply install/migrations/update_10.0.11_to_10.0.12/user.php:42]: My bad, default would be null if default is not explicitely defined.

File: src/User.php [hunk 7]
```diff
@@ -1185,6 +1198,14 @@ public function post_updateItem($history = true)
                 true
             );
         }
+
+        // Hash user_dn if is updated
```
Comment (cedric-anne) [inline src/User.php:1202]: This should be done in `pre_updateInDB` to prevent having to run an extra update process.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
You should add a test to ensure that `user_dn_hash` is correctly emptied when `user_dn` is changed to an empty value.
